telegraf/plugins/plugins.go

```diff
@@ -0,0 +1,1599 @@
+	Name        string `json:"name,omitempty"`        // Name of the plugin.
+	Description string `json:"description,omitempty"` // Description of the plugin.
+	Config      string `json:"config,omitempty"`      // Config contains the toml config of the plugin.
+}
+
+// TelegrafPlugins defines a Telegraf version's collection of plugins.
+type TelegrafPlugins struct {
+	Version string   `json:"version,omitempty"` // Version of telegraf plugins are for.
+	OS      string   `json:"os,omitempty"`      // OS the plugins apply to.
+	Plugins []Plugin `json:"plugins,omitempty"` // Plugins this version of telegraf supports.
+}
+
+// ListAvailablePlugins lists available plugins based on type.
+func ListAvailablePlugins(t string) (*TelegrafPlugins, error) {
+	switch t {
+	case "input":
+		return AvailableInputs()
+	case "output":
+		return AvailableOutputs()
+	case "processor":
+		return AvailableProcessors()
+	case "aggregator":
+		return AvailableAggregators()
+	case "bundle":
+		return AvailableBundles()
+	default:
+		return nil, fmt.Errorf("unknown plugin type '%s'", t)
+	}
+}
+
+// GetPlugin returns the plugin's sample config, if available.
+func GetPlugin(t, name string) (*Plugin, bool) {
+	var p *TelegrafPlugins
+	var err error
+
+	switch t {
+	case "input":
+		p, err = AvailableInputs()
+
+	case "output":
+		p, err = AvailableOutputs()
+
+	case "processor":
+		p, err = AvailableProcessors()
+
+	case "aggregator":
+		p, err = AvailableAggregators()
+
+	case "bundle":
+		p, err = AvailableBundles()
+
+	default:
+		return nil, false
+	}
+
+	if err != nil {
+		return nil, false
+	}
+
+	return p.findPluginByName(name)
+}
+
+// findPluginByName returns a plugin named "name". This should only be run on
+// TelegrafPlugins containing the same type of plugin.
+func (t *TelegrafPlugins) findPluginByName(name string) (*Plugin, bool) {
+	for i := range t.Plugins {
+		if t.Plugins[i].Name == name {
+			return &t.Plugins[i], true
+		}
+	}
+
+	return nil, false
+}
+
+// AvailablePlugins returns the base list of available plugins.
+func AvailablePlugins() (*TelegrafPlugins, error) {
+	all := &TelegrafPlugins{}
+
+	t, err := AvailableInputs()
+	if err != nil {
+		return nil, err
+	}
+	all.Version = t.Version
+	all.Plugins = append(all.Plugins, t.Plugins...)
+
+	t, err = AvailableOutputs()
+	if err != nil {
+		return nil, err
+	}
+	all.Plugins = append(all.Plugins, t.Plugins...)
+
+	t, err = AvailableProcessors()
+	if err != nil {
+		return nil, err
+	}
+	all.Plugins = append(all.Plugins, t.Plugins...)
+
+	t, err = AvailableAggregators()
+	if err != nil {
+		return nil, err
+	}
+	all.Plugins = append(all.Plugins, t.Plugins...)
+
+	return all, nil
+}
+
+// AvailableInputs returns the base list of available input plugins.
+func AvailableInputs() (*TelegrafPlugins, error) {
+	t := &TelegrafPlugins{}
+	err := json.Unmarshal([]byte(availableInputs), t)
+	if err != nil {
+		return nil, err
+	}
+	return t, nil
```

Comment:
nit: can we do a sort.Slice on these plugins funcs so they are ordered by name or something? to take away the juggling order in the strings below?